docs/developer/architecture/solutions.rst

```diff
@@ -0,0 +1,388 @@
+* Enforce practices (Operator pattern).
+* Guidelines for applications developers.
+
+User Stories
+------------
+
+Application import
+~~~~~~~~~~~~~~~~~~
+
+As a cluster administrator, I want to be able to import an application archive
+using a CLI tool, to make the application available for deployment.
+
+Application deployment and lifecycle
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As an application administrator, I want to manage the deployment and lifecycle
+(upgrade/downgrade/scaling/configuration/deletion) of an application using
+either a UI or through simple CLI commands (both should be available).
+
+Multiple instances of an application
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As an application administrator, I want to be able to deploy both a test
+and a prod environments for an application, without collision between them,
+so that I can qualify/test the application on the test environment.
+
+Application development
+~~~~~~~~~~~~~~~~~~~~~~~
+
+As a developer, I want to have guidelines to follow to develop an application.
+
+Application packaging
+~~~~~~~~~~~~~~~~~~~~~
+
+As a developer, I want to have documentation to know how to package an
+application.
+
+Application validation
+~~~~~~~~~~~~~~~~~~~~~~
+
+As a developer, I want to be able to know that a packaged application
+follows the requirements and is valid using a CLI tool.
+
+Design Choices
+--------------
+
+Solutions
+~~~~~~~~~
+
+What's a Solution
+"""""""""""""""""
+
+It's a packaged Kubernetes application, archived as an ISO disk image,
+containing:
+
+ * A set of OCI images to inject in MetalK8s image registry
+ * An Operator to deploy on the cluster
+ * A UI to manage and monitor the application (optional)
+
+Solution Configuration
+""""""""""""""""""""""
+
+MetalK8s already uses a ``BootstrapConfiguration`` object, stored in
+``/etc/metalk8s/bootstrap.yaml``, to define how the cluster should be
+configured from the bootstrap node, and what versions of MetalK8s are
+available to the cluster.
+
+In the same way, we will use a ``SolutionsConfiguration`` object, stored in
+``/etc/metalk8s/solutions.yaml``, to declare which Solutions are available
+to the cluster, from the bootstrap node.
+
+Here is how it will look::
+
+    apiVersion: solutions.metalk8s.scality.com/v1alpha1
+    kind: SolutionsConfiguration
+    archives:
+      - /path/to/solution/archive.iso
+    active:
+      solution-name: X.Y.Z-suffix (or 'latest')
+
+In this configuration file, no explicit information about the contents of
+archives should appear. When read by Salt at import time, the archive metadata
+will be discovered from the archive itself using a well-known file
+(``/product.txt``), which needs to define at least these 2 lines::
+
+    NAME=<solution-name>
+    VERSION=<solution-ersion>
+
+It will also be possible to define specific configuration for a Solution, using
+a ``config.yaml`` file at the root of the archive, with the following format::
```

Comment:
OK, but that's a bit redundant with the `product.txt`, no? Maybe it'd make sense to only request the more flexible YAML one (and also do it for our own ISO)?

Reply:
Yes, I agree but that does not reflect what we have now, If we want to share this architecture document with Solutions developers, it'll be confusing.
Maybe we can add a todo with the linked issue as suggested below and do this in iteration 2.

Reply:
#2422

Reply:
Arch docs are not developer docs - we are not really "sharing" them for people to act on, just to explain and describe our features and how we **plan** to implement them. Shortcuts (like a partial implementation or currently incomplete) could also be documented here, but it's more for us than for the outside.
On this aspect, I think you could have put less details in this document for the exact implementation (like format of pillar/configmaps...). Now that they are here, let's keep them, but just wanted to point that out.

But yeah, another ticket is perfectly fine to track this topic, as we don't know exactly what action to take yet.